packages/flutter/lib/src/rendering/box.dart

```diff
@@ -2360,42 +2558,105 @@ abstract class RenderBox extends RenderObject {
+              'differs from the baseline location computed by computeDryBaseline.'
+            ),
+            ErrorDescription(
+              'The $methodReturnedNull method returned null while the $methodReturnedNonNull returned a non-null $baseline of ${dryBaseline ?? realBaseline}. '
+              'Did you forget to implement $methodReturnedNull for ${objectRuntimeType(this, 'RenderBox')}?'
+            ),
+            ...messages,
+          ]);
+        } else {
+          throw FlutterError.fromParts(<DiagnosticsNode>[
+            ErrorSummary(
+              'The $baseline location returned by ${objectRuntimeType(this, 'RenderBox')}.computeDistanceToActualBaseline '
+              'differs from the baseline location computed by computeDryBaseline.'
+            ),
+            DiagnosticsProperty<RenderObject>(
+              'The RenderBox was',
+              this,
+            ),
+            ErrorDescription(
+              'The computeDryBaseline method returned $dryBaseline,\n'
+              'while the computeDistanceToActualBaseline method returned $realBaseline.\n'
+              'Consider checking the implementations of:\n'
+              ' * computeDistanceToActualBaseline\n'
+              ' * computeDryBaseline\n'
+              ' * performLayout\n'
+              'methods of the ${objectRuntimeType(this, 'RenderBox')} class and make sure they are consistent.'
+            ),
+            ...messages,
+          ]);
+        }
+      }
       return true;
-    }
-    return false;
+    }());
   }
 
   @override
   void markNeedsLayout() {
-    if (_clearCachedData() && parent is RenderObject) {
+    // If `_layoutCacheStorage.clear` returns true, then this [RenderBox]'s layout
+    // is used by the parent's layout algorithm (it's possible that the parent
+    // only used the intrinsics for paint, but there's no good way to detect that
+    // so we conservatively assume it's a layout dependency).
+    //
+    // A render object's performLayout implementation may depend on the baseline
+    // location or the intrinsic dimensions of a descendant, even when there are
+    // relayout boundaries between them. The `_layoutCacheStorage` being non-empty
+    // indicates that the parent depended on this RenderBox's baseline location,
+    // or intrinsic sizes, and thus may need relayout, regardless of relayout
+    // boundaries.
+    //
+    // Some calculations may fail (dry baseline, for example). The layout
+    // dependency is still established, but only from the RenderBox that failed
+    // to compute the dry baseline to the ancestor that queried the dry baseline.
+    if (_layoutCacheStorage.clear() && parent != null) {
+      _layoutCacheStorage.clear();
       markParentNeedsLayout();
       return;
     }
     super.markNeedsLayout();
   }
 
-  @override
-  void layout(Constraints constraints, {bool parentUsesSize = false}) {
```

Comment:
For my own understanding: Is it correct that we no longer need to clear the cache here because the cache is constraints aware now?

Reply:
right, the dry version and the wet version of the baseline calculation now share the same cache. We can't do the same for `size` because `performLayout` may also compute the child offsets while `computeDryLayout` only computes the size.